crates/shielded_token/src/masp/shielded_sync/dispatcher.rs

```diff
@@ -492,25 +508,37 @@ where
             })
             .chain(fvks.iter().copied())
         {
-            if let Some(h) = self.ctx.vk_heights.entry(vk.key).or_default() {
-                let birthday = IndexedTx::entire_block(vk.birthday);
-                if birthday > h.indexed_tx {
-                    h.indexed_tx = birthday;
-                }
-            } else if vk.birthday >= BlockHeight::first() {
-                self.ctx.vk_heights.insert(
-                    vk.key,
-                    Some(MaspIndexedTx {
-                        indexed_tx: IndexedTx::entire_block(vk.birthday),
-                        kind: MaspTxKind::Transfer,
-                    }),
+            // NB: store the viewing keys in the wallet
+            let decrypted_notes = self.ctx.pos_map.entry(vk.key).or_default();
+
+            // NB: sanity check to confirm we haven't decrypted notes
+            // before the supplied birthday
+            if vk.birthday > self.ctx.synced_height
+                && !decrypted_notes.is_empty()
+            {
+                eyre::bail!(
+                    "Invalid viewing key birthday, set after notes have \
+                     already been decrypted"
                 );
             }
+
+            // NB: store the birthday in order to potentially
+            // save some work during trial decryptions
+            self.birthdays.insert(vk.key, vk.birthday);
+
+            // NB: the min birthday will allow skipping a bunch
+            // of blocks if we are syncing from scratch
+            if vk.birthday < min_birthday {
+                min_birthday = vk.birthday
+            }
         }
 
-        // the latest block height which has been added to the witness Merkle
-        // tree
-        let last_witnessed_tx = self.ctx.note_index.keys().max().cloned();
+        if self.ctx.synced_height == BlockHeight(0)
+            && self.client.capabilities().may_fetch_pre_built_tree()
+            && !self.birthdays.is_empty()
+        {
+            self.ctx.synced_height = min_birthday;
+        }
```

Comment:
this change is kinda sketchy. if for whatever reason we finish first with the local approach, we could end up with an invalid commitment tree that's missing a bunch of data (every note committed before `min_birthday`).

might need to think of an approach to safely keep this optimization, or remove it altogether